lib/internal/util/diff.js

```diff
@@ -0,0 +1,204 @@
+      if (actual[i] !== expected[i]) {
+        // Skip the indicator for the first 2 characters because the diff is immediately apparent
+        // It is 3 instead of 2 to account for the quotes
+        if (i >= 3) {
+          indicatorIdx = i;
+        }
+        break;
+      }
+    }
+
+    if (indicatorIdx !== -1) {
+      message += `\n${StringPrototypeRepeat(' ', indicatorIdx + 2)}^`;
+    }
+  }
+
+  return { message };
+}
+
+function getSimpleDiff(originalActual, actual, originalExpected, expected) {
+  const kMaxShortStringLength = 12;
+
+  let stringsLen = actual.length + expected.length;
+  // Accounting for the quotes wrapping strings
+  if (typeof originalActual === 'string') {
+    stringsLen -= 2;
+  }
+  if (typeof originalExpected === 'string') {
+    stringsLen -= 2;
+  }
+  if (stringsLen <= kMaxShortStringLength && (originalActual !== 0 || originalExpected !== 0)) {
+    return { message: `${actual} !== ${expected}`, header: '' };
+  }
+
+  const isStringComparison = typeof originalActual === 'string' && typeof originalExpected === 'string';
+
+  if (isStringComparison && colors.hasColors) {
+    return getColoredMyersDiff(actual, expected);
+  }
+
+  return getStackedDiff(actual, expected);
+}
+
+function isSimpleDiff(actual, inspectedActual, expected, inspectedExpected) {
+  if (inspectedActual.length > 1 || inspectedExpected.length > 1) {
+    return false;
+  }
+
+  return typeof actual !== 'object' || actual === null || typeof expected !== 'object' || expected === null;
+}
+
+function getErrorMessage(operator, message) {
+  if (!operator && !message) {
+    return '';
+  }
+  return message || kReadableOperator[operator];
+}
+
+/**
+ * Generate a difference report between two values
+ * @param {any} actual - The first value to compare
+ * @param {any} expected - The second value to compare
+ * @returns {string|null} - Formatted diff string or null if values are equal
+ */
+// The following parameters are not documented because they are only used internally
+// @param {string} [operator] - The operator used in the assertion
+// @param {string} [customMessage] - Custom message to display in the diff, still just in the assertion
+function diff(actual, expected, operator, customMessage) {
+  // The equality check should only be done when not called from the assertion module
+  // When coming from there, we know for sure actual and expected are not equal
+  if (!operator) {
+    // First check if the values are deeply equal
+    try {
+      const assert = getLazy(() => require('assert'))();
+      // eslint-disable-next-line
+      assert.deepStrictEqual(actual, expected);
+      // If we reach here, the values are equal, so return null
+      return null;
+    } catch {
+      // Values are different, continue to generate diff
+    }
+  }
+
+  let skipped = false;
+  let message = '';
+  const inspectedActual = inspectValue(actual);
+  const inspectedExpected = inspectValue(expected);
```

Comment:
As external utility function, I would expect to receive already formatted input either as array or string. The way the inspection is happening here is opinionated for the assert output and would not be suitable for a generic purpose diffing functionality. Instead, users should just do the formatting on their end. They will in fact likely already have the input and formatting it here would just add extra quotes.